topdown/eval.go

```diff
@@ -1659,6 +1668,62 @@ func (e evalVirtualPartial) evalOneRule(iter unifyIterator, rule *ast.Rule, cach
 	return nil
 }
 
+func (e evalVirtualPartial) partialEvalSupport(iter unifyIterator) error {
+
+	path := e.plugged[:e.pos+1].Insert(e.e.saveNamespace, 1)
+
+	if !e.e.saveSupport.Exists(path) {
+		for i := range e.ir.Rules {
+			err := e.partialEvalSupportRule(iter, e.ir.Rules[i], path)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	rewritten := ast.NewTerm(e.ref.Insert(e.e.saveNamespace, 1))
+	return e.e.saveUnify(rewritten, e.rterm, e.bindings, e.rbindings, iter)
+}
+
+func (e evalVirtualPartial) partialEvalSupportRule(iter unifyIterator, rule *ast.Rule, path ast.Ref) error {
+
+	child := e.e.child(rule.Body)
+	child.traceEnter(rule)
+
+	e.e.saveStack.PushQuery(nil)
```

Comment:
out of curiosity, why push nil here? later on in the `child.eval` func we pop and expect a non-nil (no check), do other places normally need to check for nil when popping things off the saveStack?

Reply:
This just pushes a new empty query on to the save stack. We want expressions to be saved to a new query on the stack because the query is ultimately going to be used as a support rule body. We don't have to worry about the nil checks later on because queries are just slices of expressions and callers deal with them by checking for emptiness.